R/plotScree.R

```diff
@@ -0,0 +1,169 @@
+#' }
+#'
+#' @details
+#' \code{plotScree} creates a scree plot or eigenvalues plot, which is useful
+#' for visualizing the relative importance of components in dimensionality
+#' reduction techniques like PCA, RDA, or CCA. When the input is a
+#' TreeSummarizedExperiment, the function extracts eigenvalues from the specified
+#' reduced dimension slot. When the input is a vector, it directly uses these
+#' values as eigenvalues.
+#' 
+#' The plot can include a combination of barplot, points, connecting lines,
+#' and labels, which can be controlled using the \code{show.*} parameters.
+#' 
+#' An option to show cumulative explained variance is also available by setting
+#' \code{cumulative = TRUE}.
+#' 
+#' @return 
+#' A \code{ggplot2} object 
+#'
+#' @name plotScree
+#'
+#' @examples
+#' # Load necessary libraries
+#' library(ggplot2)
+#' 
+#' # Load dataset
+#' library(miaViz)
+#' data("enterotype", package = "mia")
+#' tse <- enterotype
+#'  
+#' # Run RDA and store results into TreeSE
+#' tse <- addRDA(
+#'     tse,
+#'     formula = assay ~ ClinicalStatus + Gender + Age,
+#'     FUN = getDissimilarity,
+#'     distance = "bray",
+#'     na.action = na.exclude
+#'     )
+#' 
+#' # Plot scree plot
+#' plotScree(tse, "RDA")
+#' 
+NULL
+
+#' @rdname plotScree
+#' @export
+setGeneric("plotScree", signature = c("x"),
+    function(x, ...) standardGeneric("plotScree"))
+
+#' @rdname plotScree
+#' @export
+setMethod("plotScree", signature = c(x = "SingleCellExperiment"),
+    function(x, dimred, cumulative = FALSE, ...) {
+        # Check if dimred exists
+        if (!dimred %in% reducedDimNames(x)) {
+            stop("'dimred' must specify a valid reducedDim.", call. = FALSE)
+        }
+        # Get reducedDim
+        reduced_dim <- reducedDim(x, dimred)
+      
+        # Extract eigenvalues
+        # Check if data is available
+        ind <- names(attributes(reduced_dim)) %in% c("eig", "varExplained")
+        if( any(ind) ){
+            # Add explained variance
+            eig <- attributes(reduced_dim)[ind][[1]]
+        } else{
+            stop("No eigenvalues found in the specified reducedDim.", 
+                 call. = FALSE)
+        }
+      
+        # Call the vector method
+        plotScree(as.numeric(eig), names(eig), cumulative = cumulative, ...)
+    }
+)
+
+#' @rdname plotScree
+#' @export
+setMethod("plotScree", signature = c(x = "vector"),
+    function(x, names = NULL, cumulative = FALSE, ...) {
+        # Ensure 'x' is numeric
+        if (!is.numeric(x)) {
+            stop("'x' must be a numeric vector.", call. = FALSE)
+        }
+        plot_data <- .prepare_data(x, names, cumulative)
+        # plot vector
+        .scree_plotter(plot_data, cumulative = cumulative, ...)
+    }
+)
+
+.prepare_data <- function(x, names = NULL, cumulative = FALSE) {
+    df <- data.frame(
+        Component = if (!is.null(names)) names else seq_along(x),
+        Eigenvalue = x
+    )
+    
```

Comment:
Remove extra line breaks, this is so short function so they do not improve readability